### PR TITLE
socket-util: make sure accept_flush() doesn't hang on EOPNOTSUPP

### LINKED ISSUE #12335: systemd v242 may go into loop when ListenDatagram socket units are stopped.
When we upgrade systemd from v241 to v242 on our systems, they  can no longer be shut down cleany, because systemd goes into an endless loop when it receives syslog data on the socket of the stopping `rsyslog.socket` unit. The problem originates in PR #12271. After commit 4ff9bc2ea6e5daf08b5c9de17c0dcf16c3b7379e ("tree-wide: port users over to use new ERRNO_IS_ACCEPT_AGAIN() call"), flush_accept() calls poll() and accept() in a loop as long as accept() returns EOPNOTSUPP, which is assumed to be a transient error [1]. This is done regardless of the socket type. However, accept() on SOCK_DGRAM sockets returns EOPNOTSUPP persistently.

Early Debug Shell session in the blocked state after  `shutdown -h now`  (manually re-typed from a vncviewer window, so typos are possible):
```
sh-4.3# uname -r
4.19.19.mx64.244
sh-4.3# strace -p 1
poll([{fd=41, events=POLLIN}],1 ,0)    = 1 ([{fd=41, revents=POLLIN}])
accept4(41, NULL, NULL, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EOPNOTSUPP (Operation not supported)
poll([{fd=41, events=POLLIN}],1 ,0)    = 1 ([{fd=41, revents=POLLIN}])
accept4(41, NULL, NULL, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EOPNOTSUPP (Operation not supported)
poll([{fd=41, events=POLLIN}],1 ,0)    = 1 ([{fd=41, revents=POLLIN}])
accept4(41, NULL, NULL, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EOPNOTSUPP (Operation not supported)
[...]
^C
sh-4.3# lsof -p 1
[...]
systemd   1 root   41u     unix 0x0000000085ff6bd2      0t0      16686 /dev/log type=DGRAM
[...]
sh-4.3:# gdb -p 1 /bin/systemd
[...]
(gdb) where
#0  0x................ in accept4 (fd=41, addr=..., addr_len=0x0, flags=526336) at ../sysdeps/unix/sysv/linux/accept4.c:32
#1  0x................ in flush_accept () from /lib/systemd/libsystemd-shared-242.so
#2  0x................ in socket_enter_running ()
#3  0x................ in socket_dispatch_io ()
#4  0x................ in source_dispatch () from /lib/systemd/libsystemd-shared-242.so
#5  0x................ in sd_event_dispatch () /lib/systemd/libsystemd-shared-242.so
#6  0x................ in sd_event_run () /lib/systemd/libsystemd-shared-242.so
#7  0x................ in manager_loop ()
#8  0x................ in main ()
```
[1] https://github.com/systemd/systemd/blob/1e5d2d656420d0e755dbcf72aeba3c3aba54e956/src/basic/socket-util.c#L1222

### PR DESCRIPTION
**Pull request by poettering**
Fixes #12335

### CODE REVIEW

**Comment by buczek** [comment]
I've tested (the first commit only) and verified, that #12335, as observed here, is indeed fixed by it.

**Comment by buczek** [comment]
What is the reason to use a local loop with the complexity of error classification instead of just returning from flush_accept and flush_fd on the first error? In the unlikely case, that there are more connects or writes pending after the first error, wouldn't that already be handled by the main loop calling into socket_enter_running again? If this comes to late, the related service might get reactivated, but this can happen based on timing anyway. We might need to consume one event so that a defective service can't be reactivated on the same event over and over, but why do we need to consume everything?

Strictly read, the accept(2) manpage (which I wouldn't trust to be complete in this regard anyway) enumerates transient errors for TCP/IP only.

**Comment by poettering** [comment]
> What is the reason to use a local loop with the complexity of error classification instead of just returning from flush_accept and flush_fd on the first error? 

The goal of flush_ports() is really to make sockets that are busy non-busy anymore. During shutdown we really want to get rid of all activation requests in full. Yes, we could let the main event loop deal with spurious accept()s, but I think this is less than ideal, since we have ratelimits and stuff on it. It should normally be fully safe to simply flush out everything queued the way we do if it wasn't for the ambiguous error code. And I am kinda confident we can make this safe even if we loop tightly.

> In the unlikely case, that there are more connects or writes pending after the first error, wouldn't that already be handled by the main loop calling into socket_enter_running again? If this comes to late, the related service might get reactivated, but this can happen based on timing anyway. We might need to consume one event so that a defective service can't be reactivated on the same event over and over, but why do we need to consume everything?
> 
> Strictly read, the accept(2) manpage (which I wouldn't trust to be complete in this regard anyway) enumerates transient errors for TCP/IP only.

Well, if we don#t list enough errors than we'd leave the loop earlier than necessary, but that's entirely OK (as that's the behaviour you are asking for anyway).

File: src/basic/socket-util.c [hunk 1]
```diff
@@ -1225,9 +1225,22 @@ int flush_accept(int fd) {
                 .fd = fd,
                 .events = POLLIN,
         };
-        int r;
+        int r, b;
+        socklen_t l = sizeof(b);
+
+        /* Similar to flush_fd() but flushes all incoming connection by accepting them and immediately
```
Comment (keszybz) [inline src/basic/socket-util.c:1231]: connection**s**

File: src/basic/socket-util.c [hunk 2]
```diff
@@ -1225,9 +1225,22 @@ int flush_accept(int fd) {
...
+         * closing them.  */
+
+        if (getsockopt(fd, SOL_SOCKET, SO_ACCEPTCONN, &b, &l) < 0)
+                return -errno;
 
-        /* Similar to flush_fd() but flushes all incoming connection by accepting them and immediately closing them. */
+        assert(l == sizeof(b));
+        if (!b) /* Let's check if this is a socket accepting connections before calling accept(). That's
+                 * because accept4() can return EOPNOTSUPP in the fd we are called on is not a listening
```
Comment (keszybz) [inline src/basic/socket-util.c:1239]: in the fd we are called on → if the fd

**Comment by paulmenzel** [comment]
On 04/23/2019 03:03 PM, Zbigniew Jędrzejewski-Szmek wrote:
> keszybz approved this pull request.
> 
> LGTM. There are some typos, but not important enough to matter.

Why didn’t you wait for some more hours so that the commit could be
amended? Now somebody else will create a merge/pull request. :(

**Comment by keszybz** [comment]
Somebody has already done that, and someone else has already merged it ;) See #12367.

**Pull request merged**